burn-ndarray/src/ops/conv.rs

```diff
@@ -25,59 +25,119 @@ pub(crate) fn conv2d<E: FloatNdArrayElement>(
 
-            for ic in (in_channels * g)..(in_channels * (g + 1)) {
-                for kh in 0..kernel_height {
-                    for kw in 0..kernel_width {
-                        for oh in 0..out_height {
-                            for ow in 0..out_width {
-                                let ih = oh * stride_height + kh * dilatation_height;
-                                let iw = ow * stride_width + kw * dilatation_width;
+                        let x = x.slice(s![b, ic, .., ..]);
+                        let k = weights.slice(s![oc, weight_ic, .., ..]);
 
-                                let weight_ic = ic - (g * in_channels);
-                                output[[b, oc, oh, ow]] +=
-                                    x[[b, ic, ih, iw]] * weight.array[[oc, weight_ic, kh, kw]];
+                        for kh in 0..kernel_height {
+                            // let kr = &k.row(kh).to_slice().unwrap()[0..kernel_width];
+                            for kw in 0..kernel_width {
+                                let k = k[[kh, kw]];
+
+                                // NOTE: This loop is duplicated twice so that the compiler can perform auto-vectorization
+                                // in the case that the stride/dilation is 1.
+                                if (1, 1, 1, 1)
+                                    == (
+                                        stride_width,
+                                        stride_height,
+                                        dilation_width,
+                                        dilation_height,
+                                    )
+                                {
+                                    for oh in 0..out_height {
+                                        // Construct a sub-slice view of the input row.
+                                        // This is done upfront so that rustc does not have to emit bounds checks
+                                        // in the hot loop below.
+                                        let ir = x.row(oh + kh).to_slice().unwrap();
+
+                                        let mut or = output.row_mut(oh);
+                                        let or = &mut or.as_slice_mut().unwrap()[0..out_width];
+
+                                        #[allow(clippy::needless_range_loop)]
+                                        for ow in 0..out_width {
+                                            let iw = ow + kw;
+                                            or[ow] += ir[iw] * k;
+                                        }
+                                    }
+                                } else {
+                                    for oh in 0..out_height {
+                                        // Construct a sub-slice view of the input row.
+                                        // This is done upfront so that rustc does not have to emit bounds checks
+                                        // in the hot loop below.
```

Comment:
I would put those comments above the two duplicated loops explaining that we are building input row (ir) and output row (or) using slices to avoid bounds checks in the hot loops.